### PR TITLE
Prohibit Java deserialisation in the Corda process

### PR DESCRIPTION
**Pull request by andr3ej**
This PR implements https://r3-cev.atlassian.net/browse/CORDA-322

### CODE REVIEW

File: core/build.gradle [hunk 1]
```diff
@@ -42,6 +42,7 @@ dependencies {
 
     // SLF4J: commons-logging bindings for a SLF4J back end
     compile "org.slf4j:jcl-over-slf4j:$slf4j_version"
+    compile "org.slf4j:jul-to-slf4j:$slf4j_version"
```
Comment (chrisr3) [inline core/build.gradle]: Doesn't this dependency belong in the Node? Because you install it in `node/src/main/kotlin/net/corda/node/Corda.kt`, and I _really_ hope that we're not using JUL anywhere else.

File: node/src/main/kotlin/net/corda/node/shell/InteractiveShell.kt [hunk 2]
```diff
@@ -80,8 +80,6 @@ object InteractiveShell {
         this.node = node
         var runSSH = runSSHServer
 
-        Logger.getLogger("").level = Level.OFF   // TODO: Is this really needed?
```
Comment (chrisr3) [inline node/src/main/kotlin/net/corda/node/shell/InteractiveShell.kt]: You can remove the `import` statements too.

File: node/src/main/kotlin/net/corda/node/Corda.kt [hunk 3]
```diff
@@ -155,15 +162,33 @@ fun main(args: Array<String>) {
...
+    println("""
 You are using a version of Java that is not supported (${System.getProperty("java.version")}). Please upgrade to the latest version.
 Corda will now exit...""")
-        exitProcess(1)
+    exitProcess(1)
+}
+
+private fun disableJavaDeserialization() {
+    // ObjectInputFilter and friends are in java.io in Java 9 but sun.misc in backports, so we use the system property interface for portability:
+    System.setProperty("jdk.serialFilter", "maxbytes=0")
```
Comment (chrisr3) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: So is our minimum JDK requirement now 8u121? (See [JDK-8155760](https://bugs.openjdk.java.net/browse/JDK-8155760))

File: core/build.gradle (hunk 1 above)
Comment (andr3ej) [inline core/build.gradle]: I only put it there because the other slf4j stuff is in there too, will move to node.

File: node/src/main/kotlin/net/corda/node/Corda.kt [hunk 4]
```diff
@@ -155,15 +162,33 @@ fun main(args: Array<String>) {
...
+    // Attempt a deserialization so that ObjectInputFilter (permanently) inits itself:
+    val data = ByteArrayOutputStream().apply { ObjectOutputStream(this).use { it.writeObject(object : Serializable {}) } }.toByteArray()
+    try {
+        // Turn down logging so users don't see REJECTED this one time:
+        withLevel("java.io.serialization", "WARN") { ObjectInputStream(data.inputStream()).use { it.readObject() } }
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: It'll be clear with the code block was on the next line. Similarly for the `data` line above.

File: node/src/main/kotlin/net/corda/node/Corda.kt (hunk 3 above)
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: Yes I believe that's the first update of 8 that has this feature.

File: node/src/main/kotlin/net/corda/node/driver/Driver.kt [hunk 5]
```diff
@@ -122,6 +122,11 @@ data class NodeHandle(
     fun rpcClientToNode(): CordaRPCClient = CordaRPCClient(configuration.rpcAddress!!)
 }
 
+data class WebserverHandle(
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/driver/Driver.kt]: This change seems to be unrelated to the task at hand. Perhaps it should be in a separate PR?

File: node/src/main/kotlin/net/corda/node/services/transactions/RaftUniquenessProvider.kt [hunk 6]
```diff
@@ -69,7 +72,20 @@ class RaftUniquenessProvider(
         val address = Address(myAddress.host, myAddress.port)
         val storage = buildStorage(storagePath)
         val transport = buildTransport(config)
-        val serializer = Serializer()
+        val serializer = Serializer().apply {
+            register(DistributedImmutableMap.Commands.PutAll::class.java) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/transactions/RaftUniquenessProvider.kt]: It's not clear what's going on here - a comment will help

File: node/src/main/kotlin/net/corda/node/services/transactions/RaftUniquenessProvider.kt [hunk 7]
```diff
@@ -69,7 +72,20 @@ class RaftUniquenessProvider(
...
+                object : TypeSerializer<DistributedImmutableMap.Commands.PutAll<*, *>> {
+                    override fun write(obj: DistributedImmutableMap.Commands.PutAll<*, *>?, buffer: BufferOutput<out BufferOutput<*>>?, serializer: Serializer?) = writeMap(obj!!.entries, buffer!!, serializer!!)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/transactions/RaftUniquenessProvider.kt]: The `!!` is only need b/c the IDE pessimistically makes the method params nullable. You can simple remove the `?`s

File: node/src/main/kotlin/net/corda/node/Corda.kt (hunk 3 above)
Comment (chrisr3) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: See `node/capsule/build.gradle`:
```
capsuleManifest {
    ...
    minJavaVersion = '1.8.0'
    minUpdateVersion['1.8'] = '102'
   ...
}
```
I think these will need updating accordingly.

Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: Yep. May as well make it the latest release whilst we're at it to get the other bugfixes.

File: node/src/main/kotlin/net/corda/node/driver/Driver.kt (hunk 5 above)
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/driver/Driver.kt]: Needed due to the change to addressMustBeBound, which itself was needed as integration tests were hanging instead of failing when a node did exitProcess(1) due to java being too old.

**Comment by shamsasari** [comment]
`ReferencesAwareJavaSerializer` can be removed I guess.

File: core/src/main/kotlin/net/corda/core/utilities/Logging.kt [hunk 8]
```diff
@@ -60,4 +61,19 @@ object LogHelper {
         config.addLogger(name, loggerConfig)
         loggerContext.updateLoggers(config)
     }
+
+    /**
+     * May fail to restore the original level due to unavoidable race if called by multiple threads.
+     */
+    fun <T> withLevel(logName: String, levelName: String, block: () -> T) = let {
```
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/utilities/Logging.kt]: The let returns the block result rather than Unit, otherwise you have to declare T and add a return, so I'm using let for brevity. Or I could get rid of the T facility altogether as I'm not using it, but propagating the block result feels like the default thing to do.

File: node/src/main/kotlin/net/corda/node/Corda.kt (hunk 3 above)
Comment (chrisr3) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: And I've just remembered that there's `web-server/webcapsule/build.gradle` too, which has a similiar JDK version check.

File: node/src/integration-test/kotlin/net/corda/node/BootTests.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+                assertEquals("filter status: REJECTED", e.message)
+            }
+        }
+    }
+
+}
+
+class ObjectInputStreamFlow : FlowLogic<Unit>() {
+
+    override fun call() {
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: What happens if the flow attempts to clear the `jdk.serialFilter`? Perhaps do that here to make it doesn't work.

Comment (chrisr3) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: You mean hostile flows can rewrite the Node's system properties? Yikes!

Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: `System.clearProperty`

**Comment by mikehearn** [comment]
Guys - flows are trusted, in the current threat model. The only parts of a cordapp that aren't trusted are the smart contract logic and state objects. So we don't have to worry about that. We might want to sandbox flows later to guide developers away from doing dumb things but we aren't intending to stop malicious flow developers (because the app has been installed by the node admin already!)

File: node/src/integration-test/kotlin/net/corda/node/BootTests.kt (hunk 9 above)
Comment (andr3ej) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: The flow would indeed be able to hack it if the node didn't consume the property up-front, I've added clearProperty to the test to enforce that. @chrisr3 Agreed, I'm allergic to system properties but in this case it seems to be the only way to avoid having separate java 8/9 builds.

File: node/src/main/kotlin/net/corda/node/Corda.kt (hunk 3 above)
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/Corda.kt]: Yes, I'll use it in the "Corda requires Java 8" message too.

**Comment by shamsasari** [comment]
Is this ready to review again?

**Comment by andr3ej** [comment]
Perfect timing, I just pushed the last fixes.
